### PR TITLE
Fix/unhandled exception in groq streaming requests

### LINKED ISSUE #1072: Unhandled exception in groq requests when stream_options field used
### What Happened?

When `stream` and `stream_options.include_usage` are used for `groq` requests, the application fails due to an unhandled exception and is terminated. I will share the curl example, response chunk comparison with groq's API, and the root cause below.

curl
```
curl --location 'http://localhost:8787/v1/chat/completions' \
--header 'Content-Type: application/json' \
--header 'x-portkey-provider: groq' \
--header 'x-portkey-custom-host: https://api.groq.com/openai/v1' \
--header 'Authorization: ****' \
--data '{
    "messages": [
        {
            "role": "user",
            "content": "hey, count to 7 with a rhyme please"
        }
    ],
    "model": "llama-3.3-70b-versatile",
    "stream": true,
    "stream_options": {
        "include_usage": true
    }
}'
```

Here are the last 3 chunks for the same request from groq

```
data: {
    "id": "chatcmpl-e7ebbb57-834f-48fa-bc73-5abb960c78b5",
    "object": "chat.completion.chunk",
    "created": 1745935006,
    "model": "llama-3.3-70b-versatile",
    "system_fingerprint": "fp_3f3b593e33",
    "choices": [
        {
            "index": 0,
            "delta": {},
            "logprobs": null,
            "finish_reason": "stop"
        }
    ],
    "x_groq": {
        "id": "req_01jt0xby4ffanamrs6m40mv0fy",
        "usage": {
            "queue_time": 0.091527015,
            "prompt_tokens": 45,
            "prompt_time": 0.003758819,
            "completion_tokens": 54,
            "completion_time": 0.196363636,
            "total_tokens": 99,
            "total_time": 0.200122455
        }
    }
}

data: {
    "id": "chatcmpl-e7ebbb57-834f-48fa-bc73-5abb960c78b5",
    "object": "chat.completion.chunk",
    "created": 1745935006,
    "model": "llama-3.3-70b-versatile",
    "system_fingerprint": "fp_3f3b593e33",
    "choices": [],
    "usage": {
        "queue_time": 0.091527015,
        "prompt_tokens": 45,
        "prompt_time": 0.003758819,
        "completion_tokens": 54,
        "completion_time": 0.196363636,
        "total_tokens": 99,
        "total_time": 0.200122455
    }
}

data: [DONE]
```

And last 3 chunks from portkey gateway

```
data: {
    "id": "chatcmpl-82aadf3e-a8d8-4a12-bf53-36306322f02f",
    "object": "chat.completion.chunk",
    "created": 1745936996,
    "model": "llama-3.3-70b-versatile",
    "provider": "groq",
    "choices": [
        {
            "index": 0,
            "delta": {
                "role": "assistant",
                "content": " tricks"
            },
            "logprobs": null,
            "finish_reason": null
        }
    ]
}

data: {
    "id": "chatcmpl-82aadf3e-a8d8-4a12-bf53-36306322f02f",
    "object": "chat.completion.chunk",
    "created": 1745936996,
    "model": "llama-3.3-70b-versatile",
    "provider": "groq",
    "choices": [
        {
            "index": 0,
            "delta": {
                "role": "assistant",
                "content": "."
            },
            "logprobs": null,
            "finish_reason": null
        }
    ]
}

data: {
    "id": "chatcmpl-82aadf3e-a8d8-4a12-bf53-36306322f02f",
    "object": "chat.completion.chunk",
    "created": 1745936996,
    "model": "llama-3.3-70b-versatile",
    "provider": "groq",
    "choices": [
        {
            "index": 0,
            "delta": {},
            "logprobs": null,
            "finish_reason": "stop"
        }
    ],
    "usage": {
        "prompt_tokens": 45,
        "completion_tokens": 53,
        "total_tokens": 98
    }
}
```

As you can see streaming is not completed with a `data: [DONE]`. The problem is that the gateway fails because of the last chunk of groq's response, which has choices field empty array. Except this last chunk, all the chunks' choice field has an item but last chunk does not have any. Here gateway fails and does not return the last DONE chunk

### What Should Have Happened?

When `groq` provider selected, `stream` and `stream_options.include_usage` field are `true` in the completion request; 

- Gateway should be able to process all the chunks, whether the choices field is present, empty or null 
- Gateway should finish streaming the request with a DONE chunk
- The gateway should not fail

### Relevant Code Snippet

`/src/providers/groq/chatComplete.ts`

```
return `data: ${JSON.stringify({
    id: parsedChunk.id,
    object: parsedChunk.object,
    created: parsedChunk.created,
    model: parsedChunk.model,
    provider: GROQ,
    choices: [
      {
        index: parsedChunk.choices[0].index || 0,
        delta: {
          role: 'assistant',
          content: parsedChunk.choices[0].delta.content,
          tool_calls: parsedChunk.choices[0].delta?.tool_calls,
        },
        logprobs: null,
        finish_reason: parsedChunk.choices[0].finish_reason || null,
      },
    ],
  })}\n\n`;
```

### Your Twitter/LinkedIn

https://www.linkedin.com/in/metinogurlu

### PR DESCRIPTION
**Pull request by metinogurlu**
Code Quality bug fix 

## Author Description

This PR fixes the [#1072](https://github.com/Portkey-AI/gateway/issues/1072) and enable the gateway to process groq completion requests when `stream_options` is used.

### 🔄 What Changed
This PR fixes an unhandled exception in Groq streaming requests by adding proper null checks and handling undefined values in the stream response. The changes focus on the `GroqChatCompleteStreamChunkTransform` function to properly handle edge cases when processing stream chunks from Groq API.

### 🔍 Impact of the Change
The fix enables the gateway to properly process Groq completion requests when `stream_options` is used, resolving issue #1072. This prevents crashes and improves reliability when streaming responses from Groq.

### 📁 Total Files Changed
1 file modified: `src/providers/groq/chatComplete.ts` with 34 additions and 21 deletions.

### 🧪 Test Added
N/A - No tests were added in this PR.

### 🔒 Security Vulnerabilities
N/A - No security vulnerabilities were addressed in this PR.

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)

## How Has This Been Tested?
- [ ] Unit Tests
- [ ] Integration Tests
- [x] Manual Testing

## Screenshots (if applicable)
N/A

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Related Issues
Fixes [#1072](https://github.com/Portkey-AI/gateway/issues/1072)

## Quality Recommendations

1. Add unit tests to verify the fix for streaming requests

2. Consider adding more comprehensive error handling for other potential edge cases in the stream response

## Sequence Diagram

```mermaid
sequenceDiagram
    participant Client
    participant Gateway
    participant GroqAPI
    
    Client->>Gateway: Send streaming chat completion request
    Gateway->>GroqAPI: Forward request with stream_options
    GroqAPI-->>Gateway: Return stream chunks
    
    loop For each stream chunk
        Gateway->>Gateway: GroqChatCompleteStreamChunkTransform()
        Note over Gateway: Handle potential undefined values
        Note over Gateway: Add proper null checks for choices
        Note over Gateway: Map usage data correctly
    end
    
    Gateway-->>Client: Return transformed stream chunks
```

### CODE REVIEW

**Review by narengogi — APPROVED** [review]

**Comment by narengogi** [comment]
LGTM! 🚀🚀🚀🚀

File: src/providers/groq/chatComplete.ts [hunk 1]
```diff
@@ -115,17 +118,27 @@ export const GroqChatCompleteStreamChunkTransform: (
...
+        ? [
+            {
+              index: parsedChunk.choices[0].index || 0,
+              delta: {
+                role: 'assistant',
+                content: parsedChunk.choices[0].delta?.content || '',
+                tool_calls: parsedChunk.choices[0].delta?.tool_calls || [],
+              },
+              logprobs: null,
+              finish_reason: parsedChunk.choices[0].finish_reason || null,
```
Comment (unsync) [inline src/providers/groq/chatComplete.ts:132]: @metinogurlu if i may offer another possible _and opinionated_ way to fix the issue and avoid returning an empty array of choices (not sure how clients can react to this, as choices are usually filled with at least one item):

```
    choices: [
      {
        index: parsedChunk.choices?.[0]?.index || 0,
        delta: {
          role: 'assistant',
          content: parsedChunk.choices?.[0].delta?.content || '',
          ...(parsedChunk.choices?.[0].delta?.tool_calls && {
            tool_calls: parsedChunk.choices?.[0].delta?.tool_calls,
          }),
        },
        logprobs: null,
        finish_reason: parsedChunk.choices?.[0]?.finish_reason || null,
      },
    ],
```

Using `choices?.[0]` makes the chaining null-safe, and the spread operator for tool_calls is there to avoid adding it when it is not in the response

Comment (metinogurlu) [reply src/providers/groq/chatComplete.ts:132]: Sure, can change the implementation accordingly but I'd like to make sure because I considered groq api response as a reference. Groq returns the choice array empty in the last chunk. Here is the example below.

```
data: {
    "id": "chatcmpl-e7ebbb57-834f-48fa-bc73-5abb960c78b5",
    "object": "chat.completion.chunk",
    "created": 1745935006,
    "model": "llama-3.3-70b-versatile",
    "system_fingerprint": "fp_3f3b593e33",
    "choices": [],
    "usage": {
        "queue_time": 0.091527015,
        "prompt_tokens": 45,
        "prompt_time": 0.003758819,
        "completion_tokens": 54,
        "completion_time": 0.196363636,
        "total_tokens": 99,
        "total_time": 0.200122455
    }
}
```

Also checked gateway's response for openapi, same empty array there as well.

```
data: {
    "id": "chatcmpl-BRfC4nDhvtPApkuN5bWYLY7ygMFuB",
    "object": "chat.completion.chunk",
    "created": 1745932904,
    "model": "gpt-3.5-turbo-0125",
    "service_tier": "default",
    "system_fingerprint": null,
    "choices": [],
    "usage": {
        "prompt_tokens": 17,
        "completion_tokens": 33,
        "total_tokens": 50,
        "prompt_tokens_details": {
            "cached_tokens": 0,
            "audio_tokens": 0
        },
        "completion_tokens_details": {
            "reasoning_tokens": 0,
            "audio_tokens": 0,
            "accepted_prediction_tokens": 0,
            "rejected_prediction_tokens": 0
        }
    }
}
```

Comment (unsync) [reply src/providers/groq/chatComplete.ts:132]: @metinogurlu fair point !
I was using other providers as reference for this choice, but I guess that for streaming responses this is less relevant that non-stream responses 👍
